src/MessageList/MessageListItem.vala

```diff
@@ -101,21 +101,27 @@ public class Mail.MessageListItem : Gtk.ListBoxRow {
         var separator = new Gtk.Separator (Gtk.Orientation.HORIZONTAL);
         separator.hexpand = true;
 
+        var infobar = new Gtk.InfoBar ();
+        infobar.margin = 12;
+        infobar.message_type = Gtk.MessageType.WARNING;
+        infobar.add_button (_("Show Images"), 1);
+        infobar.add_button (_("Always Show from Sender"), 2);
+        infobar.get_content_area ().add (new Gtk.Label (_("This message contains remote images.")));
+        infobar.get_style_context ().add_class (Gtk.STYLE_CLASS_FRAME);
+
+        ((Gtk.Box) infobar.get_action_area ()).orientation = Gtk.Orientation.VERTICAL;
+
         settings = new GLib.Settings ("io.elementary.mail");
 
         web_view = new Mail.WebView ();
-        web_view.margin = 6;
-        web_view.image_load_blocked.connect (() => {
```

Comment:
Should probably use this signal to determine when to show the infobar rather than checking the whitelist again.